sphinx/ext/apidoc/_extension.py

```diff
@@ -0,0 +1,231 @@
+        # TODO per-module automodule_options
+        automodule_options = ['members', 'undoc-members', 'show-inheritance']
+        if (_options := _check_list_of_strings(i, options, 'automodule_options')) is not None:
+            automodule_options = _options
+
+        diff = (
+            set(options)
+            - {'path', 'destination', 'exclude_patterns', 'automodule_options', 'maxdepth'}
+            - set(bool_keys)
+        )
+        if diff:
+            logger.warning(
+                __('apidoc_modules item %i has unexpected keys: %s'),
+                i,
+                ', '.join(diff),
+                type=WARNING_TYPE,
+            )
+
+        args = ExtensionOptions(
+            module_path=str(module_path),
+            destdir=str(dest_path),
+            exclude_pattern=exclude_patterns,
+            automodule_options=automodule_options,
+            maxdepth=maxdepth,
+            quiet=True,
+            **extra_options,
+        )
+
+        written_files, modules = recurse_tree(
+            args.module_path, exclude_patterns_compiled, args, args.templatedir
+        )
+        if args.tocfile:
+            written_files.append(
+                create_modules_toc_file(modules, args, args.tocfile, args.templatedir)
+            )
+        if args.remove_old:
+            _remove_old_files(written_files, args.destdir, args.suffix)
+
+
+def _check_string(index: int, options: dict[str, Any], key: str, required: bool) -> str | None:
+    """Check that a key's value is a string in the options.
+
+    :returns: the value of the key, or None if missing or it is not a string
+    """
+    if key not in options:
+        if required:
+            logger.warning(
+                __("apidoc_modules item %i must have a '%s' key"),
+                index,
+                key,
+                type=WARNING_TYPE,
+            )
+        return None
+    if not isinstance(options[key], str):
+        logger.warning(
+            __("apidoc_modules item %i '%s' must be a string"), index, type=WARNING_TYPE
+        )
+        return None
+    return options[key]
+
+
+def _check_list_of_strings(index: int, options: dict[str, Any], key: str) -> list[str] | None:
+    """Check that a key's value is a list of strings in the options.
+
+    :returns: the value of the key, or None if missing or it is not a string list
+    """
+    if key not in options:
+        return None
+    if not isinstance(options[key], list):
+        logger.warning(
+            __("apidoc_modules item %i '%s' must be a list"),
+            index,
+            key,
+            type=WARNING_TYPE,
+        )
+        return None
+    for item in options[key]:
+        if not isinstance(item, str):
+            logger.warning(
+                __("apidoc_modules item %i '%s' must contain strings"),
+                index,
+                key,
+                type=WARNING_TYPE,
+            )
+            return None
+    return options[key]
+
+
+@dataclass
```

Comment:
Perhaps?:
```suggestion
@dataclass(frozen=True)
```

I'd also maybe set ``slots=True``, but that is only Python 3.10+

Reply:
According to our release plan, we should have dropped 3.9 some months ago so maybe we could first move to 8.x by dropping 3.9 (there are quite a lot of annoying things in between 3.9 and 3.10).